### DISCUSSION TITLE
program only a hex file with "west flash"

### DISCUSSION THREAD

**Discussion by niklasparnesten** [zephyrproject-rtos/zephyr#43401] [Q&A]
Hi,

Is it possible to just program a hex-file with "west flash" , without the build dir?

In our CI we have the build stage separated from the test stage: the binaries are built on a linux server and the test are executed on a PC in our lab. So in the lab we only want to program the binaries and run our tests.

Is this possible with west or do we have to go directly on nrfjprog/openocd commands?

Best Regards,
Niklas

**Comment by DerekSnell** [comment]
Hi @niklasparnesten ,

Yes, west can flash hex files.  If you use the help command below, West prints how to point to a hex file:

`west flash --help

usage: west flash [-h] [-d DIR] [-r RUNNER] [--skip-rebuild] [-H] [--board-dir DIR] [--elf-file FILE]
                  [--hex-file FILE] [--bin-file FILE] [--gdb GDB] [--openocd OPENOCD] [--openocd-search DIR]

Permanently reprogram a board's flash with a new binary.

optional arguments:
  -h, --help            show this help message and exit

general options:
  If the build directory is not given, the default is build/ unless the
  build.dir-fmt configuration variable is set. The current directory is
  checked after that. If either is a Zephyr build directory, it is used.

  -d DIR, --build-dir DIR
                        application build directory
  -r RUNNER, --runner RUNNER
                        override default runner from --build-dir
  --skip-rebuild        do not refresh cmake dependencies first

runner configuration:
  ===================================================================
    IMPORTANT:
    Individual runners support additional options not printed here.
  ===================================================================

  Run "west flash --context" for runner-specific options.

  If a build directory is found, --context also prints per-runner
  settings found in that build directory's runners.yaml file.

  Use "west flash --context -r RUNNER" to limit output to a
  specific RUNNER.

  Some runner settings also can be overridden with options like
  --hex-file. However, this depends on the runner: not all runners
  respect --elf-file / --hex-file / --bin-file, nor use gdb or openocd,
  etc.

  -H, --context         print runner- and build-specific help
  --board-dir DIR       board directory
  --elf-file FILE       path to zephyr.elf
  --hex-file FILE       path to zephyr.hex
  --bin-file FILE       path to zephyr.bin
  --gdb GDB             path to GDB
  --openocd OPENOCD     path to openocd
  --openocd-search DIR  path to add to openocd search path, if applicable`

**Reply by niklasparnesten** [reply]
Hi @DerekSnell,

Thanks for your answer!

I have tried out several combinations of the options but cannot find what I need:
The best would be if i could omit the build folder all together and just provide the hex-file.

Seems like the build folder holds some meta-data generated on the build-machine that may not correspond to the test-machine.

**Reply by avinashpatnaik** [reply]
@niklasparnesten Have you figured out how to do this?

**Reply by nordicjm** [reply]
It is not supported to just flash a hex file outside of a build at the present time

**Reply by avinashpatnaik** [reply]
Thanks for confirming. I found an alternate way to achieve this using nrfjprog (see below) but I am not sure if this works. It is a multi-image build that I am trying to flash. Can you confirm? 

nrfjprog -f NRF53 --coprocessor CP_NETWORK --program zephyr/zephyr.hex --chiperase
nrfjprog -f NRF53 --program zephyr/zephyr.hex --chiperase

**Reply by nordicjm** [reply]
No this will not work, are you using zephyr or NCS? The hex files for the network and application cores are in different folders

**Reply by prayassamriya** [reply]
So there is no west option available to flash a pre-generated hex? that really seems strange. Can anyone share what is the dependency with build directory?

**Reply by nordicjm** [reply]
You can't just flash a hex file because the build directory has settings e.g. what flashing tool is being used, what the actual chip being used is, other tool options like SWD speed, etc.

**Reply by AndreyDodonov-EH** [reply]
Closest you can do is `west flash --build-dir  [YOUR_DIR] --runner [YOUR_RUNNER, e.g. jlink, probe-rs etc.] --file [YOUR-HEX-FILE] --no-rebuild`
